### PR TITLE
Improve Git efficiency by avoiding unnecessary fetches

### PR DESCRIPTION
**Pull request by nwalters512**
This PR aims to improve the efficiency of Git operations when saving changes made via in-browser editors. Specifically, it tries to avoid unnecessary `git fetch ...` operations.

Our hypothesis is that when someone makes a change via the in-browser editor, they're most often working with an already up-to-date version of the remote repo (that is, both repos have the same commits on the course's branch). In this case, the initial `git fetch ...` won't pull in any new changes, but we still spend ~500ms executing it because it has to talk to GitHub. So, we'll default to _not_ pulling and instead assume we can commit and push directly. If the push fails, we'll revert all changes, fetch the latest commit, re-commit the new changes on top of that, and push again.

This will make the average case 50% faster, and the worst case 50% slower. We think this is an acceptable tradeoff.

### CODE REVIEW

File: apps/prairielearn/src/lib/editors.ts [hunk 1]
```diff
@@ -223,13 +223,39 @@ export class Editor {
...
+
+            job.info('Fetch from remote git repository');
+            await job.exec('git', ['fetch'], {
+              cwd: this.course.path,
+              env: gitEnv,
+            });
+
+            await cleanAndResetRepository(this.course, gitEnv, job);
+
+            await writeAndCommitChanges();
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply apps/prairielearn/src/lib/editors.ts:250]: That sounds correct to me. Just to make sure I'm understanding, at the moment the `shouldEdit()` check is being done _before_ we acquire the course lock, leading to a TOCTTOU bug?

[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/lib/editors.ts:250]: Exactly. And specifically, it's being done before we (potentially) pull in changes from the remote, which could change what `shouldEdit()` returns.

File: apps/prairielearn/src/lib/editors.ts [hunk 2]
```diff
@@ -193,41 +199,62 @@ export abstract class Editor {
...
-                '-m',
-                commitMessage,
-              ],
-              {
-                cwd: this.course.path,
-                env: gitEnv,
-              },
-            );
+
+            await cleanAndResetRepository(this.course, gitEnv, job);
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/editors.ts:255]: Just to make sure I'm understanding, we are relying on the reset here to throw away our first commit that failed to push?

File: apps/prairielearn/src/lib/editors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwest1066) [reply apps/prairielearn/src/lib/editors.ts:257]: Understood. This was done in https://github.com/PrairieLearn/PrairieLearn/pull/9909 for future reference.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/lib/editors.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/lib/editors.ts:256]: Correct. I added a comment explaining this in 40f957a.

Comment (mwest1066) [reply apps/prairielearn/src/lib/editors.ts:256]: Perfect, thanks.

**Pull request merged**